source/margarita/upgrading.rst

```diff
@@ -0,0 +1,557 @@
+
+   * .coveragerc
+   * .gitignore
+   * README.rst
+   * setup.cfg
+   * .travis.yml (look at project.travis.yml)
+
+Vagrant Smoke Test
+------------------
+
+Now, we're going to create a fresh Vagrant VM just to make sure that our current repository deploys
+correctly.
+
+#. Edit ``conf/pillar/local/env.sls`` to look like this:
+
+   .. code-block:: yaml
+
+      environment: local
+      domain: margarita.example.com
+
+#. Edit ``conf/pillar/local/secrets.sls`` to look like this:
+
+   .. code-block:: yaml
+
+      secrets:
+        DB_PASSWORD: "dbPassword"
+        BROKER_PASSWORD: "brokerPassword"
+
+#. Add the following line to your laptop's ``/etc/hosts`` file::
+
+     33.33.33.10 margarita.example.com
+
+#. Make sure we're starting from a fresh VM:
+
+   .. code-block:: bash
+
+      ~/dev/myproject$ vagrant destroy
+      ~/dev/myproject$ vagrant up
+
+#. Deploy!
+
+   .. code-block:: bash
+
+      ~/dev/myproject$ fab vagrant setup_master
+      ~/dev/myproject$ fab -H 127.0.0.1:2222 vagrant setup_minion setup_minion:salt-master,web,worker,balancer,db-master,queue,cache
+      ~/dev/myproject$ fab vagrant deploy
+
+#. If that works, you should see your site at https://margarita.example.com.
+
+
+.. _update-repo:
+
+Update Repo Code
+----------------
+
+.. NOTE:: We will upgrade Salt on the staging machine. Once you have completed the upgrade process
+          and verified that it is working perfectly, you'll need to repeat the process for `updating
+          production`_. When you get to that point, our instructions will point you back here to
+          repeat the process for production. Just follow these instructions, replacing *staging*
+          with *production*, and *develop* with *master*.
+
+Once you have gotten the smoke test to work successfully, we'll need to get all of these changes
+into a branch that salt will be able to checkout on the staging server.
+
+#. Commit your changes locally.
+#. Push your changes to a feature branch (*your-feature-branch*) on Github.
+#. Update ``repo.branch`` in ``conf/pillar/staging/env.sls`` from *develop* to
+   *your-feature-branch*. Remember to change this back to its original value when this entire
+   process is successful. The default is *develop* for staging and *master* for production.
+
+
+.. _upgrade-salt:
+
+Upgrade Salt
+------------
+
+1. Fetch a copy of ``/etc/salt/minion`` from the server. We'll need which roles are currently being
+   used, so we can setup the same roles when we call ``setup_minion`` in step 5.
+
+   .. code-block:: bash
+
+      ~/dev/myproject$ scp 1.2.3.4:/etc/salt/minion old-minion.conf
+
+#. Uninstall salt. We're using the ``--force-yes`` parameter because salt packages are *held* on
+   some of our servers, so this is needed to allow uninstallation. **Make sure you are using the new
+   fabfile!**
+
+   .. code-block:: bash
+
+      ~/dev/myproject$ fab -H 1.2.3.4 staging -- sudo apt-get remove salt-master salt-minion salt-common -y --force-yes
```

Comment:
BTW, I never knew we could invoke arbitrary commands on the server using fabric this way. Handy!